xcm/xcm-executor/src/assets.rs

```diff
@@ -140,22 +153,99 @@ impl Assets {
 	/// Return the assets in `self`, but (asset-wise) of no greater value than `assets`.
 	///
 	/// Result is undefined if `assets` includes elements which match to the same asset more than once.
-	pub fn min<'a, I: Iterator<Item=&'a MultiAsset>>(&self, assets: I) -> Self {
+	///
+	/// Example:
+	///
+	/// ```
+	/// use xcm_executor::Assets;
+	/// use xcm::v0::{MultiAsset, MultiLocation};
+	/// let assets_i_have: Assets = vec![
+	/// 	MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount: 100 },
+	/// 	MultiAsset::AbstractFungible { id: vec![0], amount: 100 },
+	/// ].into();
+	/// let assets_they_want: Assets = vec![
+	/// 	MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount: 200 },
+	/// 	MultiAsset::AbstractFungible { id: vec![0], amount: 50 },
+	/// ].into();
+	///
+	/// let assets_we_can_trade: Assets = assets_i_have.min(assets_they_want.assets_iter());
+	/// assert_eq!(assets_we_can_trade.into_assets_iter().collect::<Vec<_>>(), vec![
+	/// 	MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount: 100 },
+	/// 	MultiAsset::AbstractFungible { id: vec![0], amount: 50 },
+	/// ]);
+	/// ```
+	pub fn min<'a, M: 'a + sp_std::borrow::Borrow<MultiAsset>, I: Iterator<Item=M>>(&self, assets: I) -> Self {
```

Comment:
The api for `saturating_take` expects a `Vec<MultiAssets>` while this API expects an Assets iterator.

Is there a reason these APIs are not the same?

Reply:
To be fair, I am not aware of any. How about this?

```patch
From 91e5ba56079000bbdbe75142d38c35aee166df1c Mon Sep 17 00:00:00 2001
From: Sergei Shulepov <s.pepyakin@gmail.com>
Date: Tue, 20 Oct 2020 13:01:23 +0200
Subject: [PATCH] How about this?

---
 xcm/xcm-executor/src/assets.rs | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/xcm/xcm-executor/src/assets.rs b/xcm/xcm-executor/src/assets.rs
index 7df21b4c9..8616b7d8b 100644
--- a/xcm/xcm-executor/src/assets.rs
+++ b/xcm/xcm-executor/src/assets.rs
@@ -174,7 +174,11 @@ impl Assets {
 	/// 	MultiAsset::AbstractFungible { id: vec![0], amount: 50 },
 	/// ]);
 	/// ```
-	pub fn min<'a, M: 'a + sp_std::borrow::Borrow<MultiAsset>, I: Iterator<Item=M>>(&self, assets: I) -> Self {
+	pub fn min<'a, M, I>(&self, assets: I) -> Self
+	where
+		M: 'a + sp_std::borrow::Borrow<MultiAsset>,
+		I: IntoIterator<Item = M>,
+	{
 		let mut result = Assets::default();
 		for asset in assets.into_iter() {
 			match asset.borrow() {
@@ -288,7 +292,10 @@ impl Assets {
 	/// 	MultiAsset::ConcreteFungible { id: MultiLocation::Null, amount: 100 },
 	/// ]);
 	/// ```
-	pub fn saturating_take(&mut self, assets: Vec<MultiAsset>) -> Assets {
+	pub fn saturating_take<I>(&mut self, assets: I) -> Assets
+	where
+		I: IntoIterator<Item = MultiAsset>,
+	{
 		let mut result = Assets::default();
 		for asset in assets.into_iter() {
 			match asset {
@@ -538,7 +545,7 @@ mod tests {
 
 		let taken_none = assets.saturating_take(none);
 		assert_eq!(None, taken_none.assets_iter().next());
-		let taken_all = assets.saturating_take(all.into());
+		let taken_all = assets.saturating_take(all);
 		// Everything taken
 		assert_eq!(None, assets.assets_iter().next());
 		let all_iter = taken_all.assets_iter();
@@ -610,7 +617,7 @@ mod tests {
 		assets2_vec.push(CNF(400));
 		let assets2: Assets = assets2_vec.into();
 
-		let taken = assets1.saturating_take(assets2.into());
+		let taken = assets1.saturating_take(Vec::from(assets2));
 		let taken = taken.into_assets_iter().collect::<Vec<_>>();
 		assert_eq!(taken, vec![CF(300), AF(1, 50), CNF(400)]);
 
-- 
2.28.0
```